runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java

```diff
@@ -0,0 +1,98 @@
+package org.corfudb.runtime.view;
+
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.collections.CorfuCounter;
+
+import java.util.UUID;
+
+/**
+ * Globally Unique Identity (GUID) generator that returns ids
+ * with weak comparable ordering with minimal distributed state sync.
+ * On startup increment a globally unique id via Corfu Object.
+ * Use this id with locally incrementing values to return unique ids.
+ * Higher bits are made of an UTC timestamp while lower bits carry the
+ * counter.
+ *
+ * Created by Sundar Sridharan on 5/22/19.
+ */
+@Slf4j
+public class CorfuGuidGenerator implements OrderedGuidGenerator {
+    private final static long TIMESTAMP_CORRECTION_MASK = 0xFF00000000000000L;
+    private final static int TIMESTAMP_CORRECTION_SHIFT = 56;
+    private final static long TIMESTAMP_MASK = 0x00FFffFFffFF0000L;
+    private final static int TIMESTAMP_SHIFT = 16;
+    private final static int MAX_TIMESTAMP_CORRECTION = 0xFF;
+    private final static int MAX_INSTANCE_ID = 0xFFff;
+
+    private final String GUID_STREAM_NAME = "CORFU_GUID_COUNTER_STREAM";
+
+    private CorfuCounter distributedCounter;
```

Comment:
An a SMR counter is not needed to obtain a unique number. Its cleaner to just use:

```java
        TokenResponse resp = rt.getSequencerView().next();
        ILogData ld = rt.getAddressSpaceView().read(resp.getSequence());
        instanceId = resp.getSequence();
```

Reply:
The main reason for having a separate counter instead of hole filling is to allow for a densely packed instance id, which, when packed in smaller fields will be much harder to collide.
If the instance id is 2 bytes long and we use the the token, it is quite possible, that between 2 calls, there are over 65K mutations, which means 2 instance ids will be the same. However if they are stored in a separate counter, we will need 65K clients to come up before we hit any collision, which is practically impossible.